compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/SingleAbstractMethodLowering.kt

```diff
@@ -0,0 +1,237 @@
+                        // Runnable(::function)
+                        createFunctionProxyInstance(superType, invokable)
+                    } else if (invokable is IrBlock && invokable.statements.last() is IrFunctionReference) {
+                        // Runnable { lambda }
+                        for (statement in invokable.statements.dropLast(1))
+                            +statement
+                        createFunctionProxyInstance(superType, invokable.statements.last() as IrFunctionReference)
+                    } else {
+                        // Fall back to materializing an invokable object.
+                        createObjectProxyInstance(superType, invokable)
+                    }
+                }
+            }
+
+            // Construct a class that implements the specified SAM interface and contains a bunch of fields
+            // in which the arguments to the constructor are stored:
+            //     class sam$n(private val arg0: T0, ...) { override fun method(...) = <whatever buildOverride returns> }
+            private fun implement(
+                superType: IrType,
+                parameters: List<IrType>,
+                buildOverride: IrSimpleFunction.(List<IrField>) -> IrBody
+            ): IrClass {
+                val superClass = superType.classifierOrFail.owner as IrClass
+                // The language documentation prohibits casting lambdas to classes, but if it was allowed,
+                // the `irDelegatingConstructorCall` in the constructor below would need to be modified.
+                assert(superClass.kind == ClassKind.INTERFACE) { "SAM conversion to an abstract class not allowed" }
+
+                val superClassName = superClass.fqNameSafe.pathSegments().joinToString("_") { it.toString() }
+                val subclass = buildClass {
+                    // TODO this is not the name some tests (e.g. kt11519) expect
+                    name = Name.identifier("sam\$$superClassName\$${cachedImplementations.size + localImplementations.size}")
+                    origin = JvmLoweredDeclarationOrigin.GENERATED_SAM_IMPLEMENTATION
+                }.apply {
+                    createImplicitParameterDeclarationWithWrappedDescriptor()
+                    parent = irClass
+                    // TODO convert all type parameters to upper bounds? See the kt11696 test.
+                    superTypes += superType
+                }
+
+                val fields = parameters.mapIndexed { i, fieldType ->
+                    buildField {
+                        name = Name.identifier("arg$i")
+                        type = fieldType
+                        origin = subclass.origin
+                        visibility = Visibilities.PRIVATE
+                    }.apply {
+                        parent = subclass
+                    }
+                }
+
+                subclass.declarations += fields
+                subclass.declarations += buildConstructor {
```

Comment:
Maybe consider using (and improving if needed) the new declarations `addConstructor`, `addFunction`, etc which at least remove the need to specify parent every time